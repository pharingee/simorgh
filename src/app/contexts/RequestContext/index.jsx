import React from 'react';
import { bool, node, oneOf, string } from 'prop-types';
import getStatsDestination from './getStatsDestination';
import getStatsPageIdentifier from './getStatsPageIdentifier';
import getOriginContext from './getOriginContext';
import getEnv from './getEnv';

export const RequestContext = React.createContext('default');

export const RequestContextProvider = ({
  children,
  bbcOrigin,
  id,
  isAmp,
  pageType,
  service,
  previousPath,
}) => {
  const { isUK, origin } = getOriginContext(bbcOrigin);
  const env = getEnv(origin);
  const platform = isAmp ? 'amp' : 'canonical';
  const statsDestination = getStatsDestination({
    isUK,
    env,
    service,
  });
  const statsPageIdentifier = getStatsPageIdentifier({
    pageType,
    service,
    id,
  });

  const value = {
    env,
    id,
    isUK,
    origin,
    pageType,
    platform,
    statsDestination,
    statsPageIdentifier,
    previousPath,
  };

  return (
    <RequestContext.Provider value={value}>{children}</RequestContext.Provider>
  );
};

RequestContextProvider.propTypes = {
  bbcOrigin: string,
  children: node.isRequired,
  id: string,
  isAmp: bool.isRequired,
  pageType: oneOf(['article', 'frontPage', 'media']).isRequired,
  service: string.isRequired,
  previousPath: string,
};

RequestContextProvider.defaultProps = {
  bbcOrigin: null,
  id: null,
  previousPath: null,
};
