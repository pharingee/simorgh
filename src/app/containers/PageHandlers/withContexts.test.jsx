import React from 'react';
import { render } from '@testing-library/react';
import { shouldShallowMatchSnapshot } from '../../../testHelpers';
import WithContexts from './withContexts';
import getOriginContext from '../../contexts/RequestContext/getOriginContext';
import getStatsDestination from '../../contexts/RequestContext/getStatsDestination';
import getStatsPageIdentifier from '../../contexts/RequestContext/getStatsPageIdentifier';
import * as requestContextImports from '../../contexts/RequestContext';

jest.mock('../../contexts/RequestContext/getOriginContext', () => jest.fn());

getOriginContext.mockImplementation(origin => ({
  isUK: true,
  origin,
}));

jest.mock('../../contexts/RequestContext/getStatsDestination', () => jest.fn());

getStatsDestination.mockImplementation(() => 'NEWS_PS_TEST');

jest.mock('../../contexts/RequestContext/getStatsPageIdentifier', () =>
  jest.fn(),
);

getStatsPageIdentifier.mockImplementation(
  () => 'news.articles.c0000000000o.page',
);

const dials = { mpulse: false };

describe('withContexts HOC', () => {
  const Component = () => <h1>All the contexts!!</h1>;
  const ContextsHOC = WithContexts(Component);

  const props = {
    bbcOrigin: 'https://www.bbc.com',
    id: 'c0000000000o',
    service: 'news',
    isAmp: true,
    pageType: 'article',
    dials,
  };

  shouldShallowMatchSnapshot(
    `should return all context providers`,
    <ContextsHOC {...props} />,
  );

  describe('assertions', () => {
    let requestContextSpy;
    beforeEach(() => {
      requestContextSpy = jest.spyOn(
        requestContextImports,
        'RequestContextProvider',
      );
    });

    const pageTypes = ['article', 'frontPage', 'chicken'];

    pageTypes.forEach(pageType => {
      it(`passing pageType==${pageType} should pass along`, () => {
        const fixture = {
          bbcOrigin: 'https://www.bbc.com',
          id: 'c0000000000o',
          service: 'news',
          isAmp: true,
          pageType,
          dials,
        };

        render(<ContextsHOC {...fixture} />);

        expect(requestContextSpy).toHaveBeenCalled();
        expect(requestContextSpy).toHaveBeenCalledWith(
          expect.objectContaining({
            pageType,
          }),
          {},
        );
      });
    });
  });
});
