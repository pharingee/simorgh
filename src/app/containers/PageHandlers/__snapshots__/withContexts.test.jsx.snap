// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withContexts HOC should return all context providers 1`] = `
<ToggleContextProvider>
  <DialContextProvider
    dials={
      Object {
        "mpulse": false,
      }
    }
  >
    <ServiceContextProvider
      service="news"
    >
      <RequestContextProvider
        bbcOrigin="https://www.bbc.com"
        id="c0000000000o"
        isAmp={true}
        pageType="article"
        previousPath={null}
        service="news"
      >
        <Component
          bbcOrigin="https://www.bbc.com"
          dials={
            Object {
              "mpulse": false,
            }
          }
          id="c0000000000o"
          isAmp={true}
          pageType="article"
          previousPath={null}
          service="news"
        />
      </RequestContextProvider>
    </ServiceContextProvider>
  </DialContextProvider>
</ToggleContextProvider>
`;
